File: src/scheduler.js
/**
 * src/scheduler.js — background loop that fires scheduled & recurring posts.
 *
 * Runs on a timer (never blocking the polling loop) and survives restarts:
 * jobs live in the store, so a redeploy resumes them.  Recurring jobs are
 * advanced to their next slot without drift.
 */

import { logger } from './logger.js';

const log = logger('scheduler');

export class Scheduler {
  constructor(store, publishFn, { interval = 3000 } = {}) {
    this.store = store;
    this.publishFn = publishFn;   // async (jobId, job) => void
    this.interval = interval;
    this.timer = null;
    this.running = false;
    this.stopped = false;
    this.ticks = 0;
    this.lastTick = null;
  }

  start() {
    if (this.timer) return;
    this.stopped = false;
    this.timer = setInterval(() => { void this.tick(); }, this.interval);
    if (typeof this.timer.unref === 'function') this.timer.unref();
    log.info(`scheduler started (every ${this.interval / 1000}s)`);
    void this.tick();
  }

  async tick() {
    if (this.running || this.stopped) return;
    this.running = true;
    this.ticks += 1;
    this.lastTick = Date.now() / 1000;
    try {
      for (const [jid, job] of this.store.dueScheduled()) {
        log.info(`publishing scheduled job ${jid} (repeat=${job.repeat || 'none'})`);
        try {
          await this.publishFn(jid, job);
          if (this.store.rescheduleRecurring(jid)) log.info(`job ${jid} rescheduled`);
          else this.store.finishScheduled(jid, { ok: true });
        } catch (err) {
          log.error(`scheduled job ${jid} failed: ${err?.message || err}`);
          this.store.finishScheduled(jid, { ok: false, error: String(err?.message || err) });
        }
      }
    } catch (err) {
      log.error(`scheduler loop error: ${err?.stack || err}`);   // keep running
    } finally {
      this.running = false;
    }
  }

  stop() {
    this.stopped = true;
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = null;
    }
  }
}

export default Scheduler;

File: apb/scheduler.py
"""apb.scheduler — background thread that fires scheduled posts on time."""

from __future__ import annotations

import logging
import threading
import time

log = logging.getLogger(__name__)


class Scheduler(threading.Thread):
    """Checks the store every few seconds and publishes due posts."""

    def __init__(self, store, publish_fn, interval=3.0):
        super().__init__(daemon=True, name="apb-scheduler")
        self.store = store
        self.publish_fn = publish_fn          # (job_id, job_dict) -> result dict
        self.interval = interval
        self._stop = threading.Event()

    def run(self):
        log.info("scheduler started (interval %.1fs)", self.interval)
        while not self._stop.is_set():
            try:
                for jid, job in self.store.due_scheduled():
                    log.info("publishing scheduled job %s (repeat=%s)",
                             jid, job.get("repeat", "none"))
                    try:
                        self.publish_fn(jid, job)
                        if self.store.reschedule_recurring(jid):
                            log.info("job %s rescheduled", jid)
                        else:
                            self.store.finish_scheduled(jid, ok=True)
                    except Exception as exc:
                        log.error("scheduled job %s failed: %s", jid, exc)
                        self.store.finish_scheduled(jid, ok=False, error=str(exc))
            except Exception as exc:  # keep the thread alive no matter what
                log.exception("scheduler loop error: %s", exc)
            self._stop.wait(self.interval)

    def stop(self):
        self._stop.set()
